File: sparc-backend/algorithms/bellmanFord.js
/**
 * Bellman-Ford Algorithm implementation
 * Computes shortest paths from a source node and detects negative cycles.
 * 
 * @param {Object} graph - { nodes: number, edges: Array<{source, target, weight}> }
 * @param {number} source - The source node index
 */
function bellmanFord(graph, source) {
  const { nodes, edges } = graph;
  const numNodes = Array.isArray(nodes) ? nodes.length : nodes;
  
  const distances = Array(numNodes).fill(Infinity);
  distances[source] = 0;
  const previous = Array(numNodes).fill(null);

  const startTime = process.hrtime.bigint();

  // Relax all edges (numNodes - 1) times
  for (let i = 1; i < numNodes; i++) {
    for (const edge of edges) {
      const u = edge.source;
      const v = edge.target;
      const weight = edge.weight;
      if (distances[u] !== Infinity && distances[u] + weight < distances[v]) {
        distances[v] = distances[u] + weight;
        previous[v] = u;
      }
    }
  }

  // Check for negative weight cycle
  let hasNegativeCycle = false;
  for (const edge of edges) {
    const u = edge.source;
    const v = edge.target;
    const weight = edge.weight;
    if (distances[u] !== Infinity && distances[u] + weight < distances[v]) {
      hasNegativeCycle = true;
      break;
    }
  }

  const endTime = process.hrtime.bigint();
  const timeTakenNs = Number(endTime - startTime);

  return {
    algorithm: 'BELLMAN_FORD',
    executionTimeMs: timeTakenNs / 1_000_000,
    distances,
    previous,
    nodesVisited: numNodes,
    relaxations: edges.length * numNodes,
    hasNegativeCycle
  };
}

module.exports = bellmanFord;

File: sparc-backend/BellmanFord.java
import java.util.Arrays;
import java.util.ArrayList;
import java.util.List;
import java.util.Collections;

public class BellmanFord {

    public static AlgorithmResult run(Graph graph, int source, Integer target) {
        long startTime = System.nanoTime();
        int numNodes = graph.nodes.size();

        double[] dist = new double[numNodes];
        Arrays.fill(dist, Double.POSITIVE_INFINITY);
        dist[source] = 0;

        Integer[] prev = new Integer[numNodes];
        
        int relaxations = 0;
        int nodesVisited = numNodes;
        
        List<Graph.Edge> allEdges = new ArrayList<>(graph.edges);
        if (!graph.directed) {
            for (Graph.Edge e : graph.edges) {
                allEdges.add(new Graph.Edge(e.target, e.source, e.weight));
            }
        }

        for (int i = 0; i < numNodes - 1; i++) {
            boolean updated = false;
            for (Graph.Edge edge : allEdges) {
                relaxations++;
                if (dist[edge.source] != Double.POSITIVE_INFINITY && dist[edge.source] + edge.weight < dist[edge.target]) {
                    dist[edge.target] = dist[edge.source] + edge.weight;
                    prev[edge.target] = edge.source;
                    updated = true;
                }
            }
            if (!updated) break;
        }

        long endTime = System.nanoTime();

        AlgorithmResult res = new AlgorithmResult();
        res.algorithm = "BELLMAN_FORD";
        res.executionTimeMs = (endTime - startTime) / 1_000_000.0;
        
        Object[] outDist = new Object[numNodes];
        for (int i = 0; i < numNodes; i++) outDist[i] = dist[i] == Double.POSITIVE_INFINITY ? null : (int)dist[i];
        res.distances = outDist;

        List<Integer> path = new ArrayList<>();
        if (target != null && prev[target] != null) {
            Integer curr = target;
            while (curr != null) {
                path.add(curr);
                curr = prev[curr];
            }
            Collections.reverse(path);
        } else if (target != null && source == target) {
            path.add(source);
        }
        res.path = path;
        res.nodesVisited = nodesVisited;
        res.relaxations = relaxations;

        return res;
    }
}
